### PR TITLE
model_tools: support injection of hash algorithms

### PR DESCRIPTION
**Pull request by jkamelin**
Support extension of download engine with different hash algorithms

New checksum format for model.yml
by default for sha384:
```
checksum: str
```
for other algorithms extensions:
```
checksum:
  $type: str (sha256|sha384|md5sum|…)
  value: str
```

### CODE REVIEW

File: tools/model_tools/src/openvino/model_zoo/download_engine/cache.py [hunk 1]
```diff
@@ -39,17 +41,17 @@ def __init__(self, cache_dir):
         self._staging_dir = self._cache_dir / 'staging'
         self._staging_dir.mkdir(exist_ok=True)
 
-    def _hash_path(self, hash):
-        assert len(hash) == self._HASH_LEN
+    def _hash_path(self, hash, hash_len):
+        assert len(hash) == hash_len
```
Comment (Wovchena) [inline tools/model_tools/src/openvino/model_zoo/download_engine/cache.py:45]: That's strange, just for a debug assertion you've changed the interface of three functions. May be remove the assertion?

File: tools/model_tools/src/openvino/model_zoo/download_engine/cache.py [hunk 2]
```diff
@@ -91,6 +93,37 @@ def put(self, hash, path):
...
+        sha384_str = validation.validate_string('"sha384"', checksum['value'])
+        if not cls.RE_SHA384SUM.fullmatch(sha384_str):
+            raise validation.DeserializationError(
+                '"sha384": got invalid hash {!r}'.format(sha384_str))
+
+        sha384 = bytes.fromhex(sha384_str)
+        return cls(sha384)
+
+
+Checksum.types['sha384'] = ChecksumSHA384
```
Comment (Wovchena) [inline tools/model_tools/src/openvino/model_zoo/download_engine/cache.py:124]: Sometime you call hashing algorithms `algo` and sometimes `type`. I suggest being consistent :)

**Review by Wovchena — CHANGES_REQUESTED** [review]
Just a note. It would have been easier to review, if you had split the first commit on two: .py changes and .yml changes.

File: tools/model_tools/src/openvino/model_zoo/download_engine/cache.py (hunk 1 above)
Comment (jkamelin) [reply tools/model_tools/src/openvino/model_zoo/download_engine/cache.py]: Done

File: tools/model_tools/src/openvino/model_zoo/download_engine/cache.py (hunk 2 above)
Comment (jkamelin) [reply tools/model_tools/src/openvino/model_zoo/download_engine/cache.py:123]: Done

**Review by Wovchena — APPROVED** [review]

**Comment by Wovchena** [comment]
@jkamelin, btw, why would they want to tweak hash algo?

**Comment by jkamelin** [comment]
> 
OneZoo team has problems transitioning to sha384. One of them is to support old version of models with sha256

**Comment by jkamelin** [comment]
> @jkamelin, btw, why would they want to tweak hash algo?
OneZoo team has problems transitioning to sha384. One of them is to support old version of models with sha256

**Review by vladimir-dudnik — APPROVED** [review]

**Comment by Wovchena** [comment]
Jenkins please retry a build

**Comment by Wovchena** [comment]
Jenkins please retry a build

**Pull request merged**